src/parser/type_checking.c

```diff
@@ -12982,28 +12957,193 @@ pt_character_length_for_node (PT_NODE * node, const PT_TYPE_ENUM coerce_type)
   return precision;
 }
 
+namespace Func
+{
+  #define X(id, ...) #id,
+  const char* type_str[] = {
+    #include "func_type.x"
+  };
+  #undef X
+
+  #define X(id, signatures) signatures,
+  std::vector<func_signature>* types[] = {
+    #include "func_type.x"
+  };
+  #undef X
+
+  bool cmp_types_normal(const pt_arg_type& type, pt_type_enum type_enum)
+  {
+    return (type.type == pt_arg_type::NORMAL && type.val.type == type_enum);
+  }
+
+  bool cmp_types_generic(const pt_arg_type& type, pt_type_enum type_enum)
+  {
+    return pt_are_equivalent_types(type, type_enum);
+  }
+
+  /*
+   * get_signature () - get function signature using a function to compare types
+   */
+  const func_signature* get_signature (parser_node* node, 
```

Comment:
The expression type inference algorithm has "score" based metric when computing the "best" signature.
This algorithm seems to choose first available/not rejected signature.
I guess that this assumes we are dependent on the order of signature definition.

Reply:
yes, but also the score based algorithm is order dependent because is very likely to have same score for 2 different signatures and you need to chose only one